Rules/AvoidOverwritingBuiltInCmdlets.cs

```diff
@@ -0,0 +1,284 @@
+        private bool initialized;
+
+
+        /// <summary>
+        /// Construct an object of AvoidOverwritingBuiltInCmdlets type.
+        /// </summary>
+        public AvoidOverwritingBuiltInCmdlets() : base()
+        {
+            initialized = false;
+            cmdletMap = new Dictionary<string, HashSet<string>>();
+            Enable = true;  // Enable rule by default
+            
+            string versionTest = string.Join("", PowerShellVersion);
+
+            if (versionTest != "core-6.1.0-windows" && versionTest != "desktop-5.1.14393.206-windows")
+            {
+                // PowerShellVersion is not already set to one of the acceptable defaults
+                // Try launching `pwsh -v` to see if PowerShell 6+ is installed, and use those cmdlets
+                // as a default. If 6+ is not installed this will throw an error, which when caught will
+                // allow us to use the PowerShell 5 cmdlets as a default.
+                try
+                {
+                    var testProcess = new Process();
+                    testProcess.StartInfo.FileName = "pwsh";
+                    testProcess.StartInfo.Arguments = "-v";
+                    testProcess.StartInfo.CreateNoWindow = true;
+                    testProcess.StartInfo.UseShellExecute = false;
+                    testProcess.Start();
+                    PowerShellVersion = new[] {"core-6.1.0-windows"};
+                }
+                catch
+                {
+                    PowerShellVersion = new[] {"desktop-5.1.14393.206-windows"};
+                }
+            }
+        }
+
+
+        /// <summary>
+        /// Analyzes the given ast to find the [violation]
+        /// </summary>
+        /// <param name="ast">AST to be analyzed. This should be non-null</param>
+        /// <param name="fileName">Name of file that corresponds to the input AST.</param>
+        /// <returns>A an enumerable type containing the violations</returns>
+        public override IEnumerable<DiagnosticRecord> AnalyzeScript(Ast ast, string fileName)
+        {
+            if (ast == null)
+            {
+                throw new ArgumentNullException(nameof(ast));
+            }
+
+            var functionDefinitions = ast.FindAll(testAst => testAst is FunctionDefinitionAst, true);
+            if (functionDefinitions.Count() < 1)
+            {
+                // There are no function definitions in this AST and so it's not worth checking the rest of this rule
+                return null;
+            }
+
+            else
+            {
+                var diagnosticRecords = new List<DiagnosticRecord>();
+                if (!initialized)
+                {
+                    Initialize();
+                    if (!initialized)
+                    {
+                        throw new Exception("Failed to initialize rule " + GetName());
+                    }
+                }
+
+                foreach (var functionDef in functionDefinitions)
```

Comment:
It's not clear on this line what the type of the LHS is

```suggestion
                foreach (FunctionDefinitionAst functionDef in functionDefinitions)
```